website/addons/osfstorage/model.py

```diff
@@ -136,139 +98,111 @@ def create_waterbutler_log(self, auth, action, metadata):
         pass
 
 
-class BaseFileObject(StoredObject):
+class OsfStorageFileNode(StoredObject):
     __indices__ = [
         {
             'key_or_list': [
-                ('path', pymongo.ASCENDING),
+                ('name', pymongo.ASCENDING),
+                ('kind', pymongo.ASCENDING),
+                ('parent', pymongo.ASCENDING),
                 ('node_settings', pymongo.ASCENDING),
             ],
             'unique': True,
         }
     ]
 
-    path = fields.StringField(required=True, index=True)
-    node_settings = fields.ForeignField(
-        'OsfStorageNodeSettings',
-        required=True,
-        index=True,
-    )
-
-    _meta = {
-        'abstract': True,
-    }
-
-    @property
-    def name(self):
-        _, value = os.path.split(self.path)
-        return value
+    _id = fields.StringField(primary=True, default=lambda: str(bson.ObjectId()))
 
-    @property
-    def extension(self):
-        _, value = os.path.splitext(self.path)
-        return value
-
-    @property
-    def node(self):
-        return self.node_settings.owner
+    is_deleted = fields.BooleanField(default=False)
+    name = fields.StringField(required=True, index=True)
+    kind = fields.StringField(required=True, index=True)
+    parent = fields.ForeignField('OsfStorageFileNode', index=True)
+    versions = fields.ForeignField('OsfStorageFileVersion', list=True)
+    node_settings = fields.ForeignField('OsfStorageNodeSettings', required=True, index=True)
 
     @classmethod
-    def parent_class(cls):
-        raise NotImplementedError
+    def get(cls, path, node_settings):
+        return cls.find_one(
+            Q('_id', 'eq', path) &
+            Q('node_settings', 'eq', node_settings)
+        )
 
     @classmethod
-    def find_by_path(cls, path, node_settings):
-        """Find a record by path and root settings record.
-
-        :param str path: Path to file or directory
-        :param node_settings: Root node settings record
-        """
-        try:
-            obj = cls.find_one(
-                Q('path', 'eq', path.rstrip('/')) &
-                Q('node_settings', 'eq', node_settings._id)
-            )
-            return obj
-        except modm_errors.NoResultsFound:
-            return None
+    def get_folder(cls, path, node_settings):
```

Comment:
`path` and `_id` are now the same, effectively.
This variant of OSFStorage has the same structuring as googledrive does.
`path` is just defined as `_id` with a leading `/` and a trailing `/` if it is a folder.